### PR TITLE
[Giskard] Fixed node transition into running after parent node is done and child node not being unpaused after being paused by parent node, when child node pause condition is UNKNOWN

### PR DESCRIPTION
**Pull request by MideDay**
Added Tests for Bugs and other LifeCycle transition edge cases.

Changed how LifeCycleState compiles transitions to include parent nodes in templates.
Removed apply_goal_conditions_to_children since the new LifeCycleState transitions handle those now.

Related Issue: #11

## New Transition Summary:
Previously motion statechart templates required the parent node to apply it's conditions onto the child nodes in some for.
Now, instead of combining the child nodes condition with the parent node, the LifeCycle transitions include the parent nodes. 
This moves the logic for transitioning with motion statechart template back into the LifeCycleState and fixes a bug, where a child node can run for at least 1 tick after the parent node has already transitioned to DONE.
For example, the transition from NOT_STARTED to RUNNING now checks if the parent node's `start_condition` is fulfilled and also that it's `end_condition` is not satisfied.
Transition from PAUSED to RUNNING now includes UNKNOWN observation state. As such a node can transition from PAUSED to RUNNING if the pause_condition is UNKNOWN
Image with Life Cycle FSM without and with parent node (the parent nodes are iterative)
LifeCycleFSM

### CODE REVIEW

File: giskardpy/src/giskardpy/motion_statechart/motion_statechart.py [hunk 1]
```diff
@@ -113,51 +113,97 @@ class LifeCycleState(State):
...
+            pause_and_chain_false = create_condition(
+                node,
+                lambda p: p.pause_condition,
+                cas.TrinaryFalse,
+                cas.trinary_logic_and,
+            )
+
+            not_started_condition = node.start_condition == cas.TrinaryTrue
+            current = node
+            while current.parent_node is not None:
```
Comment (maltehue) [inline giskardpy/src/giskardpy/motion_statechart/motion_statechart.py]: can you introduce more helper methods to reduce the complexity of the code?

File: giskardpy/src/giskardpy/motion_statechart/test_nodes/test_nodes.py [hunk 2]
```diff
@@ -73,3 +77,88 @@ def build(self, context: BuildContext) -> NodeArtifacts:
...
+        self.cancel.start_condition = self.ticking1.observation_variable
+
+    def build(self, context: BuildContext) -> NodeArtifacts:
+        return NodeArtifacts(
+            observation=cas.Expression(self.ticking2.observation_variable)
+        )
+
+
+@dataclass(repr=False, eq=False)
+class TestEndBeforeStart(Goal):
```
Comment (ichumuh) [inline giskardpy/src/giskardpy/motion_statechart/test_nodes/test_nodes.py]: docs

File: giskardpy/src/giskardpy/motion_statechart/test_nodes/test_nodes.py [hunk 3]
```diff
@@ -73,3 +77,88 @@ def build(self, context: BuildContext) -> NodeArtifacts:
         return NodeArtifacts(
             observation=cas.Expression(self.inner.observation_variable)
         )
+
+
+@dataclass(repr=False, eq=False)
+class TestRunAfterStop(Goal):
```
Comment (maltehue) [inline giskardpy/src/giskardpy/motion_statechart/test_nodes/test_nodes.py]: add a docstring to the Test Goal classes. The names are descriptive already, but its easy to forget details of the transition logic after a few weeks

File: test/giskardpy_test/test_motion_statechart/test_motion_statechart.py [hunk 4]
```diff
@@ -2219,3 +2222,376 @@ def test_constraint_collection(pr2_world: World):
...
+        node2.start_condition = runafterstop.observation_variable
+
+        end = EndMotion()
+        msc.add_node(end)
+        end.start_condition = node2.observation_variable
+
+        kin_sim = Executor(world=World())
+        kin_sim.compile(motion_statechart=msc)
+        kin_sim.tick_until_end()
+        msc.draw("muh.pdf")
```
Comment (maltehue) [inline test/giskardpy_test/test_motion_statechart/test_motion_statechart.py]: disable writing of a pdf  on default. That should not happen when running the tests in the CI

File: test/giskardpy_test/test_motion_statechart/test_motion_statechart.py [hunk 5]
```diff
@@ -2219,3 +2222,376 @@ def test_constraint_collection(pr2_world: World):
...
+        node2.start_condition = runafterstopfrompause.observation_variable
+
+        end = EndMotion()
+        msc.add_node(end)
+        end.start_condition = node2.observation_variable
+
+        kin_sim = Executor(world=World())
+        kin_sim.compile(motion_statechart=msc)
+        kin_sim.tick_until_end()
+        msc.draw("muh.pdf")
```
[the remark this answers is not in the document]
Comment (MideDay) [reply test/giskardpy_test/test_motion_statechart/test_motion_statechart.py]: Added Sequences where they seemed applicable and used add_nodes in other places.

File: giskardpy/src/giskardpy/motion_statechart/motion_statechart.py
[the remark this answers is not in the document]
Comment (MideDay) [reply giskardpy/src/giskardpy/motion_statechart/motion_statechart.py]: Added Test fails on this behavior, solution is being investigated.

File: giskardpy/src/giskardpy/motion_statechart/motion_statechart.py [hunk 6]
```diff
@@ -106,6 +106,82 @@ def __eq__(self, other: Self) -> bool:
...
+            cas.trinary_logic_not(parent.end_condition),
+            parent.start_condition == cas.TrinaryTrue,
+        )
+        current = parent
+    return not_started_condition
+
+
+def _create_unpause_condition(
+    node: MotionStatechartNode,
+) -> cas.Expression:
```
Comment (ichumuh) [inline giskardpy/src/giskardpy/motion_statechart/motion_statechart.py]: if it is a free standing method and node is the only parameter, then why is it not a self method of MotionStatechartNode?

File: giskardpy/src/giskardpy/motion_statechart/motion_statechart.py [hunk 7]
```diff
@@ -106,6 +106,82 @@ def __eq__(self, other: Self) -> bool:
         return str(self) == str(other)
 
 
+def _create_condition(
+    start_from: MotionStatechartNode,
+    condition_getter: Callable[[MotionStatechartNode], cas.Expression],
+    expected_value: cas.Expression,
+    combine_func: Callable[[cas.Expression, cas.Expression], cas.Expression],
+) -> cas.Expression:
```
Comment (ichumuh) [inline giskardpy/src/giskardpy/motion_statechart/motion_statechart.py]: this method feels too complicated for me.
_create_condition is a very non descriptive name and a condition_getter as a callable attribute is difficult to understand too. i can see that you want to avoid that it depends on the specific condition. but then maybe this method should belong to TrinaryCondition?

File: test/giskardpy_test/test_motion_statechart/test_motion_statechart.py [hunk 8]
```diff
@@ -2219,3 +2223,385 @@ def test_constraint_collection(pr2_world: World):
...
+        msc = MotionStatechart()
+
+        node1 = CountTicks(ticks=1)
+        node2 = ConstTrueNode()
+        end = EndMotion()
+
+        msc.add_nodes(nodes=[node1, node2, end])
+
+        end.start_condition = node1.observation_variable
+        end.end_condition = node2.observation_variable
```
Comment (ichumuh) [inline test/giskardpy_test/test_motion_statechart/test_motion_statechart.py]: why does EndMotion have an end condition?
